### PR TITLE
Non-record: 11L mixed int5/int6 + working QAT + TTT (val_bpb=1.1466)

### PR DESCRIPTION
**Pull request by vytautas-bunevicius**
## Summary

Non-record submission stacking 8 techniques on PR #315 (1.1248):

- Working QAT fix (PR #315's was dead code due to torch.compile)
- Mixed int5 (MLP) / int6 (attention) quantization + 3% magnitude pruning
- Test-time training (3 epochs SGD post-quant, 83s on 8xH100)
- BigramHash 10240 (up from 2048)
- 64 learnable memory tokens
- Backout connection (1 scalar param)
- Per-head temperature (88 params)
- Eval stride 32

**val_bpb = 1.1466** on 8xH100 SXM. Ran with PyTorch SDPA instead of FA3 (110ms/step, 5129 steps instead of ~7000). Artifact: 14.7MB.

## Test plan

- [x] Smoke test on 1xH100 (completed)
- [x] Full run on 8xH100 SXM (completed, 605s training + 340s eval)
- [ ] Rerun with FlashAttention 3 for improved score
- [ ] 3-seed reproducibility (single seed so far)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: 11L mixed int5/int6 + working QAT + TTT (val_bpb=1.1466)

**BPB:** 1.1466 | **Compliance:** FLAG — Pre-Quant TTT runs multi-epoch on `val_tokens` with no score-first discipline

**What I found in the code** (head SHA `75168efc0eed`, file `records/track_10min_16mb/2026-03-22_11L_MixedInt56_QAT_TTT_1.1466/train_gpt.py`):

At line **1380** the pre-quant TTT function takes `val_tokens` as an input argument and runs an epoch loop over it with `loss.backward()`/`optimizer.step()`, with no prior `torch.no_grad()` scoring pass over the same tokens:

```
ttt_adapt(args, model, device, val_tokens, rank, world_size, log_fn) — for epoch in range(args.ttt_epochs), loss.backward() without prior no_grad score pass
```

**Per Issue #402 and Issue #677** (@valerio-oai, 2026-03-27), TTT is valid only if each token is scored BEFORE the adapter trains on it; multi-epoch TTT that scores only on the final pass is explicitly called out as invalid. This implementation matches the pattern that closed PR #1376 (stukenov) and was subsequently confirmed in #1485/#1487/#1488/#1489/#1517/#1539 — see [Issue #677 meta-comment from 2026-04-11](https://github.com/openai/parameter-golf/issues/677#issuecomment-4229813460) which lists the 6+ PRs in the cluster.

**Contrast with the legal Pre-Quant TTT pattern** (e.g. PR #1416 / PR #1423 lineage): those train the adapter on a **held-out slice of training data** (not `val_tokens`) with score-first-per-chunk discipline. The distinction is on the function signature itself — the argument tensor passed in.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.13s, dim=512, layers=11, vocab=1024, code=86751 B, SMOKE_TEST_PASS

**Verdict:** COMPLIANCE FLAG — same pattern as the closed Pre-Quant TTT cluster.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **CLOSE** under the same ruling as #1376 and the rest of the cluster. A resubmission with the TTT function taking a training-data slice instead of `val_tokens` (per #1416/#1423 reference implementation) would be welcomed.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.13s, dim=512, layers=11, vocab=1024, code=86751 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
